### PR TITLE
Issue #3279 - WebSocket May Hang Forever

### PR DESCRIPTION
**Pull request by joakime**
Primarily addressing the #3279 FrameFlusher hang due to CLOSE frame detection,
But also taking care of a host of other CLOSE handling as well.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

**Comment by joakime** [comment]
SNAPSHOT built of this PR/branch.
Found at https://oss.sonatype.org/content/repositories/jetty-snapshots/
Version: `9.4.15-20190205.170525-51`

**Comment by sbordet** [comment]
@joakime this became a 95 files PR (!), so let's run thought it together when you are available.

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/AbstractWebSocketConnection.java [hunk 1]
```diff
@@ -159,22 +173,112 @@ public Executor getExecutor()
...
+        }
+        String reason = cause.getMessage();
+        if (StringUtil.isBlank(reason))
+        {
+            // an exception without a message.
+            reason = cause.getClass().getSimpleName();
+        }
+
+        CloseInfo closeInfo = new CloseInfo(statusCode, reason);
+        session.callApplicationOnClose(closeInfo);
```
Comment (joakime) [inline jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/AbstractWebSocketConnection.java]: If application calls session.close() who does the disconnect?

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/AbstractWebSocketConnection.java [hunk 2]
```diff
@@ -414,63 +465,49 @@ protected void onFillInterestedFailed(Throwable cause)
...
-                LOG.debug("onReadTimeout - Connection Already CLOSED");
-            // close already completed, extra timeouts not relevant
-            // allow underlying connection and endpoint to disconnect on its own
-            return true;
+            // an exception without a message.
+            reason = timeout.getClass().getSimpleName();
         }
 
-        try
+        if (connectionState.canWriteWebSocketFrames())
```
Comment (joakime) [inline jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/AbstractWebSocketConnection.java]: Consider connectionState.timeout() instead

Comment (joakime) [reply jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/AbstractWebSocketConnection.java]: Decided to not use connectionState.timeout() but just to forward the Throwable to close(cause);

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/AbstractWebSocketConnection.java (hunk 1 above)
Comment (joakime) [reply jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/AbstractWebSocketConnection.java]: The DisconnectCallback performs the disconnect on callback failure as well.

**Comment by gregw** [comment]
@joakime @lachlan-roberts do any of these changes need to be adapted into jetty-10?

**Comment by joakime** [comment]
I'll work on the jetty-9.4.x to jetty-10.0.x merge on monday for this.

**Comment by gregw** [comment]
@lachlan-roberts can you review/help @joakime merge

**Pull request merged**